src/main/java/life/qbic/util/ProgressBar.java

```diff
@@ -1,80 +1,112 @@
+  private long start;
+  private long remainingTime;
+
+  public ProgressBar() {}
+
+  public ProgressBar(String fileName, long totalFileSize) {
+    this.fileName = shortenFileName(fileName);
+    this.totalFileSize = totalFileSize;
+    this.downloadedSize = 0L;
+    this.stepSize = totalFileSize / BARSIZE;
+    this.nextProgressJump = this.stepSize;
+    this.unitDisplay = UnitConverterFactory.determineBestUnitType(totalFileSize);
+    this.start = System.currentTimeMillis();
+    this.remainingTime = 0L;
+  }
+
+  public void updateProgress(int addDownloadedSize) {
+    this.downloadedSize += (long) addDownloadedSize;
+    checkForJump();
+  }
+
+  private String shortenFileName(String fullFileName) {
+    String shortName;
+    if (fullFileName.length() > MAXFILENAMESIZE) {
+      shortName = fullFileName.substring(0, MAXFILENAMESIZE - 3) + "...";
+    } else {
+      shortName = fullFileName;
     }
+    return shortName;
+  }
 
-    private void checkForJump() {
-        if(this.downloadedSize > this.nextProgressJump){
-            this.nextProgressJump += this.stepSize;
-            drawProgress();
-        }
+  private void checkForJump() {
+    if (this.downloadedSize > this.nextProgressJump) {
+      this.nextProgressJump += this.stepSize;
+      drawProgress();
     }
+  }
 
-    private void drawProgress() {
-        System.out.print(String.format("%-" + computeLeftPadding() +"s %s\r", this.fileName,buildProgressBar()));
-    }
+  private void drawProgress() {
+    System.out.print(
+        String.format("%-" + computeLeftPadding() + "s %s\r", this.fileName, buildProgressBar()));
+  }
 
-    private int computeLeftPadding(){
-        return MAXFILENAMESIZE + 5;
-    }
+  private int computeLeftPadding() {
+    return MAXFILENAMESIZE + 5;
+  }
 
+  private String buildProgressBar() {
+    StringBuilder progressBar = new StringBuilder("[");
+    int numberProgressStrings = Math.min((int) (this.downloadedSize / this.stepSize), BARSIZE);
 
-    private String buildProgressBar(){
-        StringBuilder progressBar = new StringBuilder("[");
-        int numberProgressStrings = Math.min((int) (this.downloadedSize / this.stepSize), BARSIZE);
+    double elapsedTimeSeconds = (System.currentTimeMillis() - this.start) / 1000.0;
 
-        for (int i = 0; i<numberProgressStrings; i++){
-            progressBar.append("#");
-        }
-        for (int i = numberProgressStrings; i<BARSIZE; i++){
-            progressBar.append(" ");
-        }
+    // Download Speed in Mbyte/s
+    double downloadSpeed = this.downloadedSize / (1000000.0 * elapsedTimeSeconds);
 
-        progressBar.append("]\t");
-        progressBar.append(String.format("%6.02f/%-7.02f%s", unitDisplay.convertBytesToUnit(this.downloadedSize),
-                unitDisplay.convertBytesToUnit(this.totalFileSize), unitDisplay.getUnitType()));
+    // Estimate remaining download time
+    long remainingDownloadTime = estimateRemainingTime(downloadSpeed * 1000000.0 / 1000.0);
```

Comment:
What are the factors here? Why multiply by 1 Millionen and divide by 1 thousand?